server/grpc/orchestrator/execution.go

```diff
@@ -0,0 +1,151 @@
+package orchestrator
+
+import (
+	"context"
+	fmt "fmt"
+
+	"github.com/mesg-foundation/engine/cosmos"
+	"github.com/mesg-foundation/engine/execution"
+	"github.com/mesg-foundation/engine/ext/xstrings"
+	"github.com/mesg-foundation/engine/hash"
+	"github.com/mesg-foundation/engine/protobuf/acknowledgement"
+	executionmodule "github.com/mesg-foundation/engine/x/execution"
+)
+
+type executionServer struct {
+	rpc  *cosmos.RPC
+	auth *Authorizer
+}
+
+// NewExecutionServer creates a new Execution Server.
+func NewExecutionServer(rpc *cosmos.RPC, auth *Authorizer) ExecutionServer {
+	return &executionServer{
+		rpc:  rpc,
+		auth: auth,
+	}
+}
+
+// Create creates an execution.
+func (s *executionServer) Create(ctx context.Context, req *ExecutionCreateRequest) (*ExecutionCreateResponse, error) {
+	// check authorization
+	if err := s.auth.IsAuthorized(ctx, req); err != nil {
+		return nil, err
+	}
+
+	// create execution
+	acc, err := s.rpc.GetAccount()
+	if err != nil {
+		return nil, err
+	}
+	eventHash, err := hash.Random()
+	if err != nil {
+		return nil, err
+	}
+	msg := executionmodule.MsgCreate{
+		Signer:       acc.GetAddress(),
+		EventHash:    eventHash,
+		ExecutorHash: req.ExecutorHash,
+		Inputs:       req.Inputs,
+		Price:        req.Price,
+		Tags:         req.Tags,
+		TaskKey:      req.TaskKey,
+	}
+	tx, err := s.rpc.BuildAndBroadcastMsg(msg)
+	if err != nil {
+		return nil, err
+	}
+	return &ExecutionCreateResponse{
+		Hash: tx.Data,
+	}, nil
+}
+
+// Stream returns stream of executions.
+func (s *executionServer) Stream(req *ExecutionStreamRequest, stream Execution_StreamServer) error {
+	// check authorization
+	if err := s.auth.IsAuthorized(stream.Context(), req); err != nil {
+		return err
+	}
+
+	// create rpc event stream
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+	subscriber := xstrings.RandASCIILetters(8)
+	query := fmt.Sprintf("%s.%s EXISTS", executionmodule.EventType, executionmodule.AttributeKeyHash)
+	eventStream, err := s.rpc.Subscribe(ctx, subscriber, query, 0)
+	defer s.rpc.Unsubscribe(context.Background(), subscriber, query)
+	if err != nil {
+		return err
+	}
+	if err := acknowledgement.SetStreamReady(stream); err != nil {
+		return err
+	}
+
+	// listen to event stream
+	for {
+		select {
+		case event := <-eventStream:
+			attrHash := fmt.Sprintf("%s.%s", executionmodule.EventType, executionmodule.AttributeKeyHash)
+			attrs := event.Events[attrHash]
+			alreadySeeHashes := make(map[string]bool)
+			for _, attr := range attrs {
+				// skip already see hash. it deduplicate same execution in multiple event.
+				if alreadySeeHashes[attr] {
+					continue
+				}
+				alreadySeeHashes[attr] = true
+				hash, err := hash.Decode(attr)
+				if err != nil {
+					return err
```

Comment:
Is it a good thing to return the error here, shouldn't we have an error channel?
Here the first error will actually stop the whole stream.
Might have the same issue on the runner API which might be more critical

Reply:
There is no error channel with gRPC streams. The only way to return the error to the client is to return it as it does.